### DISCUSSION TITLE
An Issue With Completion Items

### DISCUSSION THREAD

**Discussion by Kitosight** [microsoft/vscode-discussions#419] [Extension Development QnA] [answered]
Since I explained the problem in the other thread in my other issue, I'll just link it -> https://github.com/microsoft/vscode-discussions/discussions/414#discussioncomment-4694290.

If needed, I can provide the actual part of the code related to this, and if needed I also can show a screen recording of the actual issue.

**Comment by alefragnani** [comment, marked as answer]
Hi,

I’m sorry, but I didn’t understand the real issue you are facing. I’m not sure you are having issues with the `$` character or with fixed position.

I suggest you to provide not only a sample of the _content_ you are parsing and how/where you _expect_ the completion item to be activated, but also the piece of code of your completion item provider, and the actual behavior.

Be aware that VS Code supports multi-cursor editing, so your code must handle/ignore it, depending on your needs.

**Reply by Kitosight** [reply]
> but also the piece of code of your completion item provider, and the actual behavior.
Alright, I'll show all this very soon. I'm away from my PC right now.

**Reply by Kitosight** [reply]
So, here's a piece of code.
```js
const vscode = require('vscode');

// Shorcuts
const lang = vscode.languages;
const kind = vscode.CompletionItemKind;

function snippets() {
	const range = undefined; // For now
	
	// $addButton[]
	lang.registerCompletionItemProvider('bds', {
		provideCompletionItems() {
			const completion = new vscode.CompletionItem();
			completion.insertText = new vscode.SnippetString('\\$addButton[${1|no,yes|};${2:Button ID/URL};${3:Label};${4|primary,success,danger,secondary,link|};${5|no,yes|};${6:Emoji};${7:Message ID}]');
			completion.documentation = new vscode.MarkdownString('```bds\n$addButton[]\n```');
			completion.kind = kind.Function;
			completion.label = '$addButton';
			completion.detail = 'Adds button to the response.';
			completion.range = range;

			return [completion];
		}
	});
	// Other Completion Items
	// ...
}
```

The issue is, it doesn't match the `$`, so Auto Completion doesn't work correctly and therefore [I was told to use `Range`](https://github.com/microsoft/vscode/issues/170374) to solve this issue.
![image](https://user-images.githubusercontent.com/70456337/213263712-1d2f9ef7-cbd0-4644-ab27-fa4911d4192d.png)
But then I encountered other issue - Completion Items work only on the first character of the first line of the editor. I thought that I could retrieve the current character and line, but uh- The problem is kinda left the same.
![image](https://user-images.githubusercontent.com/70456337/213263283-d15a861f-c7b9-4dfe-aaed-12d1fb63f6a7.png)
![image](https://user-images.githubusercontent.com/70456337/213263343-69451d3e-36a8-4478-aec5-d69cf0a80aa1.png)

What I did:
```js
const vscode = require('vscode');

// Shorcuts
const lang = vscode.languages;
const kind = vscode.CompletionItemKind;
const _selection = vscode.window.activeTextEditor.selection;

function snippets() {
	const range = new vscode.Range(_selection.active.line, _selection.active.character, _selection.end.line, _selection.end.character);
	// $addButton[]
	lang.registerCompletionItemProvider('bds', {
		provideCompletionItems() {
			const completion = new vscode.CompletionItem();
			completion.insertText = new vscode.SnippetString('\\$addButton[${1|no,yes|};${2:Button ID/URL};${3:Label};${4|primary,success,danger,secondary,link|};${5|no,yes|};${6:Emoji};${7:Message ID}]');
			completion.documentation = new vscode.MarkdownString('```bds\n$addButton[]\n```');
			completion.kind = kind.Function;
			completion.label = '$addButton';
			completion.detail = 'Adds button to the response.';
			completion.range = range;

			return [completion];
		}
	});
	// Other Completion Items
	// ...
}
```
Then, I tried to use listener related to selection, but I totally fucked up and I didn't succeed.

I hope you or anyone can help me solve this issue which is killing me everyday when I'm trying to solve thus on my own.

**Reply by alefragnani** [reply]
The `range` property in `CompletionItem` is used to say _which text should be replaced_ when your completion item is selected. Which means you must detect this range, looking at the current line where the completion item was fired.

This is what the documentation says:

```ts
        /**
         * A range or a insert and replace range selecting the text that should be replaced by this completion item.
         *
         * When omitted, the range of the {@link TextDocument.getWordRangeAtPosition current word} is used as replace-range
         * and as insert-range the start of the {@link TextDocument.getWordRangeAtPosition current word} to the
         * current position is used.
         **/
```
Your code from the previous post is creating a single character range on position 0. Doing so, it will always put the completion item at position 0.

```ts
const range = new vscode.Range(new vscode.Position(0,0), new vscode.Position(0,1));
```

It seems you would like to replace _the current word that starts with `$` symbol, like `$addButton` with your completion item_. In this case, you must create a `Range` that reflects that. You must parse the line and determine the `Range`. Something like:


```ts
const r = new vscode.Range(new vscode.Position(position.line, position.character - lenghtof$addButton), new vscode.Position(position.line, position.character));
```

But, be aware that Completion Item Providers requires trigger characters, and by its documentation

```ts
        /**
         * A completion item provider can be associated with a set of `triggerCharacters`. When trigger
         * characters are being typed, completions are requested but only from providers that registered
         * the typed character. Because of that trigger characters should be different than {@link LanguageConfiguration.wordPattern word characters},
         * a common trigger character is `.` to trigger member completions.
         *
         **/
```
Also, it seems that if you provide a `range` to a `CompletionItem`, it will filter the item. To solve that, you must also provide a value to `ComplerionItem.filterText` in order to you completion item to be properly displayed

Hope this helps

**Reply by Kitosight** [reply]
> The `range` property in `CompletionItem` is used to say _which text should be replaced_ when your completion item is selected. Which means you must detect this range, looking at the current line where the completion item was fired.
> 
> This is what the documentation says:
> 
> ```ts
>         /**
>          * A range or a insert and replace range selecting the text that should be replaced by this completion item.
>          *
>          * When omitted, the range of the {@link TextDocument.getWordRangeAtPosition current word} is used as replace-range
>          * and as insert-range the start of the {@link TextDocument.getWordRangeAtPosition current word} to the
>          * current position is used.
>          **/
> ```
> Your code from the previous post is creating a single character range on position 0. Doing so, it will always put the completion item at position 0.
> 
> ```ts
> const range = new vscode.Range(new vscode.Position(0,0), new vscode.Position(0,1));
> ```
> 
> It seems you would like to replace _the current word that starts with `$` symbol, like `$addButton` with your completion item_. In this case, you must create a `Range` that reflects that. You must parse the line and determine the `Range`. Something like:
> 
> 
> ```ts
> const r = new vscode.Range(new vscode.Position(position.line, position.character - lenghtof$addButton), new vscode.Position(position.line, position.character));
> ```
> 
> But, be aware that Completion Item Providers requires trigger characters, and by its documentation
> 
> ```ts
>         /**
>          * A completion item provider can be associated with a set of `triggerCharacters`. When trigger
>          * characters are being typed, completions are requested but only from providers that registered
>          * the typed character. Because of that trigger characters should be different than {@link LanguageConfiguration.wordPattern word characters},
>          * a common trigger character is `.` to trigger member completions.
>          *
>          **/
> ```
> Also, it seems that if you provide a `range` to a `CompletionItem`, it will filter the item. To solve that, you must also provide a value to `ComplerionItem.filterText` in order to you completion item to be properly displayed
> 
> Hope this helps

I'll try out this and I'll give feedback here if I succeed or not.

*Please don't mark this discussion as completed (or resolved, or whatever *we* call it here)*

**Reply by alefragnani** [reply]
Take a look at The VS Code Extensions Sample repo for additional info https://github.com/microsoft/vscode-extension-samples/tree/main/completions-sample

**Reply by Kitosight** [reply]
> The `range` property in `CompletionItem` is used to say _which text should be replaced_ when your completion item is selected. Which means you must detect this range, looking at the current line where the completion item was fired.
> 
> This is what the documentation says:
> 
> ```ts
>         /**
>          * A range or a insert and replace range selecting the text that should be replaced by this completion item.
>          *
>          * When omitted, the range of the {@link TextDocument.getWordRangeAtPosition current word} is used as replace-range
>          * and as insert-range the start of the {@link TextDocument.getWordRangeAtPosition current word} to the
>          * current position is used.
>          **/
> ```
> Your code from the previous post is creating a single character range on position 0. Doing so, it will always put the completion item at position 0.
> 
> ```ts
> const range = new vscode.Range(new vscode.Position(0,0), new vscode.Position(0,1));
> ```
> 
> It seems you would like to replace _the current word that starts with `$` symbol, like `$addButton` with your completion item_. In this case, you must create a `Range` that reflects that. You must parse the line and determine the `Range`. Something like:
> 
> 
> ```ts
> const r = new vscode.Range(new vscode.Position(position.line, position.character - lenghtof$addButton), new vscode.Position(position.line, position.character));
> ```
> 
> But, be aware that Completion Item Providers requires trigger characters, and by its documentation
> 
> ```ts
>         /**
>          * A completion item provider can be associated with a set of `triggerCharacters`. When trigger
>          * characters are being typed, completions are requested but only from providers that registered
>          * the typed character. Because of that trigger characters should be different than {@link LanguageConfiguration.wordPattern word characters},
>          * a common trigger character is `.` to trigger member completions.
>          *
>          **/
> ```
> Also, it seems that if you provide a `range` to a `CompletionItem`, it will filter the item. To solve that, you must also provide a value to `ComplerionItem.filterText` in order to you completion item to be properly displayed
> 
> Hope this helps

Unfortunately I didn't succeed again. Also, I didn't really understand it.

*Just in case I'll note that I'll have over 300 completion items like `$addButton`.*

**Reply by Kitosight** [reply]
> Take a look at The VS Code Extensions Sample repo for additional info https://github.com/microsoft/vscode-extension-samples/tree/main/completions-sample 

I always see examples in this repo before doing something, but it can't help solve my current issue(s) in any way.
